File: python/TcpPack.py
import struct
from socket import *
import threading
from threading import RLock
import time

REV_LEN = 1024
MAX_LISTEN_NUM = 20
GUID_LEN = 32
NewMsg_Size_LEN = 4
MarkLen = 1
ConnMark_Def = 0xFF

# Server: void IServerRev(byte mark, byte[] recvBytes,   erroMsg );
# Client: void IClientRev(byte[] recvBytes,   erroMsg);
class newMsgMark:
    def __init__(self):
        super().__init__()
        self.NewMsgMarkStr = b"6DB62435625348A28426C876F6B04A88"
        self.NewMsgMarkStrEnd = b"CE1D1E7C9DBF4951BB414B6D9D2AA1C1"
    def isNewMsgMark(self,recvBuf):
        for i in range(GUID_LEN):
            if self.NewMsgMarkStr[i] != recvBuf[i]:
                 return False
        EndGDIDPos = REV_LEN - GUID_LEN
        for i in range(GUID_LEN):
            if self.NewMsgMarkStrEnd[i] != recvBuf[i + EndGDIDPos]:
                 return False
        for i in range(GUID_LEN + NewMsg_Size_LEN,EndGDIDPos):
            if 0 != recvBuf[i]:
                 return False
        return True
class Receiver:
    def __init__(self):
        self.__revSize = 0
        self.__msgSize = 0
        self.__newMsgMark =  newMsgMark()
        self.__InitBuf()
    def __InitBuf(self,bufsize=REV_LEN):
        self.__buf = bytearray()
        for i in range(bufsize):
            self.__buf.append(0)

    def rev(self,recvBuf, RevFinishFun):
        if self.__newMsgMark.isNewMsgMark(recvBuf):
            self.__revSize = 0
            self.__msgSize = struct.unpack('<I', recvBuf[GUID_LEN:GUID_LEN+NewMsg_Size_LEN])[0]
            if len(self.__buf) < self.__msgSize: self.__InitBuf(self.__msgSize)
            return
        
        CpySize = REV_LEN
        if self.__revSize + REV_LEN > self.__msgSize:
            CpySize = self.__msgSize - self.__revSize

        if 0 == self.__msgSize: return
        if len(self.__buf) < self.__revSize + CpySize: return

        for i in range(0,CpySize):self.__buf[self.__revSize+i] = recvBuf[i]
        self.__revSize = self.__revSize + CpySize
        
        if self.__revSize != self.__msgSize: return

        RevFinishFun(self.__buf, self.__revSize)
        self.__msgSize = 0
        self.__revSize = 0
class Sender:
    def __InitBuf(self):
        self.__SendBuf = bytearray()
        for i in range(REV_LEN):
            self.__SendBuf.append(0)
    def __init__(self):
        self.__InitBuf()
        self.__newMsgMark = newMsgMark()

    def __Send(self,sock):
        by = bytes(self.__SendBuf)
        sock.send(by)

    def SendMsg(self,sock, msg):
        msgSize = len(msg)
        self.__InitBuf()
        for i in range(0,GUID_LEN): self.__SendBuf[i] = self.__newMsgMark.NewMsgMarkStr[i]
        msgSizeBytes = struct.pack('<I',msgSize)
        for i in range(0,4): self.__SendBuf[i+GUID_LEN] = msgSizeBytes[i]
        pos = REV_LEN - GUID_LEN
        for i in range(0,GUID_LEN): self.__SendBuf[i + pos] = self.__newMsgMark.NewMsgMarkStrEnd[i]
        self.__Send(sock)

        hasSendSize = 0
        while True:
            self.__InitBuf()

            if msgSize <= REV_LEN:
                for i in range(0,msgSize): self.__SendBuf[i] = msg[i+hasSendSize]
                self.__Send(sock)
                return

            for i in range(0,REV_LEN): self.__SendBuf[i] = msg[i+hasSendSize]
            self.__Send(sock)

            hasSendSize = hasSendSize + REV_LEN
            msgSize = msgSize - REV_LEN
class MsgCtl:
    def __init__(self):
        self.__Receiver =  Receiver()
        self.__Sender =  Sender()
    def Rev(self,recvBuf, RevFinishFun):
        self.__Receiver.rev(recvBuf, RevFinishFun)
    def Send(self,sock, msg):
        self.__Sender.SendMsg(sock, msg)
class Connect:
    def __init__(self,_conn, _IServerRev, _IClientRev):
        self.__MsgCtl = MsgCtl()
        self.__ConnMark = ConnMark_Def
        self.__newMsgMark = newMsgMark()

        self.__IServerRev = _IServerRev
        self.__IClientRev = _IClientRev

        self.connectNormal = True
        self.__conn = _conn
        
        self.__ProcRevThr  = True
        self.__revThread = threading.Thread(target=self.__OnRev)  
        self.__revThread.setDaemon(True)  
        self.__revThread.start()
    def Close(self):
        if(None != self.__conn): self.__conn.close() 
    def Rev(self, msg,  msgSize,  erro):
        if None != msg : msg = msg[0:msgSize]
        if (None != self.__IServerRev): self.__IServerRev(self.__ConnMark, msg, erro)
        else: self.__IClientRev.Rev(msg, erro)
    def __OnRev(self):
        try:
            self.OnRev()
        except Exception as e:
            self.connectNormal = False
            self.Rev(None,0,e)
            print(self.__ConnMark)
            print(e)    

    def OnRev(self):
        recvBytes = bytearray()
        for i in range(REV_LEN):
            recvBytes.append(0)
        recBufPos = 0
        BufLen = 0

        if (self.__IServerRev != None):   
            markBytes =  self.__conn.recv(REV_LEN)
            self.__ConnMark = markBytes[0]
            print("connect mark :" + str(self.__ConnMark))

        while(True):
            if(self.__ProcRevThr == False): break 
            tempRevLen = REV_LEN - recBufPos
            tempRevBytes = self.__conn.recv(tempRevLen)            
            BufLen = len(tempRevBytes)
            for i in range(0,BufLen): recvBytes[recBufPos+i] = tempRevBytes[i]
            recBufPos += BufLen
            if recBufPos != REV_LEN: continue
            recBufPos = 0
            self.__MsgCtl.Rev(recvBytes,lambda revByte, revSize: self.Rev(revByte, revSize, ""))
    
    def Send(self,msg):
        try:
            self.__MsgCtl.Send(self.__conn, msg)
        except Exception as e:
            print(e)
            return False
        return True


class SocketServer:
    def __init__(self):
        self.__SocketConnect = []
        self.__lock = RLock()
    def Start(self,port,  _revProc):
        self.__revProc = _revProc
        self.__socketServer = socket(AF_INET, SOCK_STREAM)
        self.__socketServer .bind(("0.0.0.0",port))
        
        threadSocket = threading.Thread(target=self.__Accept)  
        threadSocket.setDaemon(True)  
        threadSocket.start()
    def __Accept(self):
        self.__socketServer .listen(MAX_LISTEN_NUM)
        while True:
            clientsock,clientaddress=self.__socketServer.accept()
            print('connect from:',clientaddress)
            self.__lock.acquire()
            self.__lock.acquire()
            self.__RemoveErroConnect()
            conn = Connect(clientsock, self.__revProc, None)
            self.__SocketConnect.append(conn)
            self.__lock.release()
            self.__lock.release()
    def __RemoveErroConnect(self):
        self.__SocketConnect = list(filter(lambda x: x.connectNormal == True,self.__SocketConnect))

    def SendToAllConn(self,msg):
        self.__SendToConn(msg)   
    def SendToMarkConn(self, mark, msg):
        self.__SendToConn(msg, mark)
    def __SendToConn(self, msg,  mark = ConnMark_Def):
        self.__lock.acquire()
        self.__lock.acquire()
        for conn in self.__SocketConnect:
            if conn.connectNormal == False: continue

            if mark != ConnMark_Def:
                if mark != conn.GetConnMark(): continue
            
            conn.Send(msg)
            
        self.__lock.release()
        self.__lock.release()

class SocketClient:
    def __init__(self):
        self.__Connect = None

    def Start(self,mark,  ip,  port, IClientRev):
        socketClient = socket(AF_INET, SOCK_STREAM)
        socketClient.connect((ip,port)) 
        socketClient.send(struct.pack('<B',mark))      
        self.__Connect = Connect(socketClient, None, IClientRev)

    def Send(self, msg):
        if None == self.__Connect: return False
        return self.__Connect.Send(msg)

File: go/packsocket/tcppack.go
package packsocket

import (
	"bytes"
	"container/list"
	"encoding/binary"
	"fmt"
	"net"
	"sync"
	"time"
)

func uint32ToBytes(data uint32) []byte {
	bytebuf := bytes.NewBuffer([]byte{})
	binary.Write(bytebuf, binary.LittleEndian, data)
	return bytebuf.Bytes()
}
func bytesToUInt32(bys []byte, startpos int) int {
	bytebuff := bytes.NewBuffer(bys[startpos : startpos+4])
	var data uint32
	binary.Read(bytebuff, binary.LittleEndian, &data)
	return int(data)
}

const (
	SOCKET_REV_BUF_SIZE = 1024

	GUID_LEN         = 16
	NewMsgMarkStr    = "6DB62435625348A28426C876F6B04A88"
	NewMsgMarkStrEnd = "CE1D1E7C9DBF4951BB414B6D9D2AA1C1"
	NewMsg_Size_LEN  = 4
	CMD_LEN          = 4

	Pos_cmd                   = 20
	Pos_msg                   = 24
	Pos_NewMsgMarkStrEnd      = 1008
	newMsgMaxDataSpace        = 984
	ConnMark_Def         byte = 255
)

//IServerRecver serverimp
type IServerRecver interface {
	OnRec(mark byte, cmd uint32, recvBuf []byte, BufLen int)
}

//IClientRecver clientimp
type IClientRecver interface {
	OnRec(cmd uint32, recvBuf []byte, BufLen int)
}

type iRevFinish interface {
	OnRevFinish(cmd uint32, recvBuf []byte, BufLen int)
}

/////////receiver
type receiver struct {
	m_revSize       int
	m_msgSize       int
	m_cmd           uint32
	m_buf           []byte
	m_NewMsgMark    []byte
	m_NewMsgMarkEnd []byte
}

func (_this *receiver) init() {
	_this.m_revSize = 0
	_this.m_msgSize = 0
	_this.m_cmd = 0
	_this.m_buf = make([]byte, SOCKET_REV_BUF_SIZE)
	_this.m_NewMsgMark = []byte(NewMsgMarkStr)
	_this.m_NewMsgMarkEnd = []byte(NewMsgMarkStrEnd)
}
func (_this *receiver) rev(recvBuf []byte, Fun iRevFinish) {
	if _this.isNewMsgMark(recvBuf) {
		_this.m_revSize = 0
		_this.m_msgSize = bytesToUInt32(recvBuf, GUID_LEN)
		_this.m_cmd = uint32(bytesToUInt32(recvBuf, Pos_cmd))
		if len(_this.m_buf) < int(_this.m_msgSize) {
			_this.m_buf = make([]byte, _this.m_msgSize)
		}

		if _this.m_msgSize <= newMsgMaxDataSpace {
			copy(_this.m_buf[0:_this.m_msgSize], recvBuf[Pos_msg:Pos_msg+_this.m_msgSize])
			_this.msgFinish(Fun)
		}
		return
	}

	CpySize := SOCKET_REV_BUF_SIZE
	if _this.m_revSize+SOCKET_REV_BUF_SIZE > _this.m_msgSize {
		CpySize = _this.m_msgSize - _this.m_revSize
	}

	copy(_this.m_buf[_this.m_revSize:], recvBuf[0:CpySize])
	_this.m_revSize = _this.m_revSize + CpySize

	if _this.m_revSize == _this.m_msgSize {
		_this.msgFinish(Fun)
	}
}
func (_this *receiver) isNewMsgMark(recvBuf []byte) bool {
	if !_this.checkGUID(recvBuf, _this.m_NewMsgMark, 0) {
		return false
	}
	if !_this.checkGUID(recvBuf, _this.m_NewMsgMarkEnd, Pos_NewMsgMarkStrEnd) {
		return false
	}
	return true
}
func (_this *receiver) msgFinish(Fun iRevFinish) {
	Fun.OnRevFinish(_this.m_cmd, _this.m_buf, _this.m_msgSize)
	_this.m_msgSize = 0
	_this.m_revSize = 0
}
func (_this *receiver) checkGUID(recvBuf []byte, guid []byte, posStart int) bool {
	for i := 0; i < GUID_LEN; i++ {
		if guid[i] != recvBuf[i+posStart] {
			return false
		}
	}
	return true
}

//////

///////////sender
type sender struct {
	m_NewMsgMark    []byte
	m_NewMsgMarkEnd []byte
}

func (_this *sender) init() {
	_this.m_NewMsgMark = []byte(NewMsgMarkStr)
	_this.m_NewMsgMarkEnd = []byte(NewMsgMarkStrEnd)
}
func (_this *sender) sendmsgFun(sock net.Conn, SendBuf []byte) bool {
	n, err := sock.Write(SendBuf)
	if err != nil {
		fmt.Println("sendmsg error:", err)
	}
	return SOCKET_REV_BUF_SIZE == n
}
func (_this *sender) SendMsg(sock net.Conn, cmd uint32, msg []byte, msgSize int) bool {
	SendBuf := make([]byte, SOCKET_REV_BUF_SIZE)

	copy(SendBuf[0:GUID_LEN], _this.m_NewMsgMark[0:GUID_LEN])
	copy(SendBuf[GUID_LEN:GUID_LEN+NewMsg_Size_LEN], uint32ToBytes(uint32(msgSize))[0:NewMsg_Size_LEN])
	copy(SendBuf[Pos_cmd:Pos_cmd+CMD_LEN], uint32ToBytes(cmd)[0:CMD_LEN])
	copy(SendBuf[Pos_NewMsgMarkStrEnd:Pos_NewMsgMarkStrEnd+GUID_LEN], _this.m_NewMsgMarkEnd[0:GUID_LEN])
	if msgSize <= newMsgMaxDataSpace {
		copy(SendBuf[Pos_msg:Pos_msg+msgSize], msg[0:msgSize])
		return _this.sendmsgFun(sock, SendBuf)
	}
	suc := _this.sendmsgFun(sock, SendBuf)
	if !suc {
		return false
	}

	hasSendSize := 0
	for {
		if msgSize <= SOCKET_REV_BUF_SIZE {
			copy(SendBuf[0:msgSize], msg[hasSendSize:hasSendSize+msgSize])
			return _this.sendmsgFun(sock, SendBuf)
		}

		copy(SendBuf[0:SOCKET_REV_BUF_SIZE], msg[hasSendSize:hasSendSize+SOCKET_REV_BUF_SIZE])
		suc = _this.sendmsgFun(sock, SendBuf)
		if !suc {
			return false
		}

		hasSendSize += SOCKET_REV_BUF_SIZE
		msgSize -= SOCKET_REV_BUF_SIZE
	}
}

////////socketConn
type socketConn struct {
	m_Conn           net.Conn
	m_pIServerRecver *IServerRecver
	m_pIClientRecver *IClientRecver

	m_sender   sender
	m_receiver receiver

	m_bConnect bool

	m_ConnMark byte
}

func (_this *socketConn) OnRevFinish(cmd uint32, recvBuf []byte, BufLen int) {
	if nil != _this.m_pIServerRecver {
		(*_this.m_pIServerRecver).OnRec(_this.m_ConnMark, cmd, recvBuf, BufLen)
	} else {
		(*_this.m_pIClientRecver).OnRec(cmd, recvBuf, BufLen)
	}
}
func (_this *socketConn) revmsg(recvBuf []byte) int {
	revlen, err := _this.m_Conn.Read(recvBuf)
	if err != nil {
		fmt.Println("revmsg error:", err)
		return 0
	}
	return revlen
}
func (_this *socketConn) init(Conn net.Conn, pIServerRecver *IServerRecver, pIClientRecver *IClientRecver) {
	_this.m_Conn = Conn
	_this.m_bConnect = true
	_this.m_ConnMark = ConnMark_Def
	_this.m_pIServerRecver = pIServerRecver
	_this.m_pIClientRecver = pIClientRecver
	_this.m_sender.init()
	_this.m_receiver.init()
	go func() {
		recvBuf := make([]byte, SOCKET_REV_BUF_SIZE)
		recBufPos := 0
		BufLen := 0

		if _this.m_pIServerRecver != nil {
			BufLen = _this.revmsg(recvBuf[:1])
			if _this.recvErro(BufLen <= 0) {
				return
			}
			_this.m_ConnMark = recvBuf[0]
		}

		var Fun iRevFinish = _this
		for {

			BufLen = _this.revmsg(recvBuf[recBufPos:])
			recBufPos += BufLen

			if _this.recvErro(BufLen <= 0) {
				return
			}

			if recBufPos != SOCKET_REV_BUF_SIZE {
				continue
			}
			recBufPos = 0

			_this.m_receiver.rev(recvBuf, Fun)
		}
	}()
}
func (_this *socketConn) recvErro(bErro bool) bool {
	if !bErro {
		return false
	}
	_this.m_bConnect = false
	_this.OnRevFinish(0, nil, 0)
	_this.m_Conn.Close()
	return true
}
func (_this *socketConn) Send(cmd uint32, msg []byte, msgSize int) bool {
	if !_this.m_bConnect {
		return false
	}
	return _this.m_sender.SendMsg(_this.m_Conn, cmd, msg, msgSize)
}

////////

///////ServerSocket
type ServerSocket struct {
	m_sockSrv        net.Listener
	m_pISocketRecver *IServerRecver

	m_Connect *list.List
	m_lock    *sync.Mutex
}

func (_this *ServerSocket) Start(port string, pISocketRecver *IServerRecver) bool {
	_this.m_lock = new(sync.Mutex)

	_this.m_Connect = list.New()

	_this.m_pISocketRecver = pISocketRecver

	l, err := net.Listen("tcp", ":"+port)
	if err != nil {
		fmt.Println("listen error:", err)
		return false
	}

	go func() {
		for {
			conn, err := l.Accept()
			if err != nil {
				fmt.Println("accept error:", err)
				continue
			}

			_this.m_lock.Lock()
			_this.removeInvalid()
			connect := new(socketConn)
			_this.m_Connect.PushBack(connect)
			connect.init(conn, _this.m_pISocketRecver, nil)
			_this.m_lock.Unlock()
		}
	}()

	return true
}
func (_this *ServerSocket) SendToAllConn(cmd uint32, msg []byte, msgSize int) bool {
	return _this.sendToConn(cmd, msg, msgSize, ConnMark_Def)
}
func (_this *ServerSocket) SendToMarkConn(mark byte, cmd uint32, msg []byte, msgSize int) bool {
	return _this.sendToConn(cmd, msg, msgSize, mark)
}
func (_this *ServerSocket) removeInvalid() {
	temp := list.New()
	for i := _this.m_Connect.Front(); i != nil; i = i.Next() {
		var conn *socketConn = i.Value.(*socketConn)
		if conn.m_bConnect {
			temp.PushBack(i)
		}
	}
	_this.m_Connect = temp
}
func (_this *ServerSocket) sendToConn(cmd uint32, msg []byte, msgSize int, mark byte) bool {
	sendSuc := false
	_this.m_lock.Lock()

	for i := _this.m_Connect.Front(); i != nil; i = i.Next() {
		var conn *socketConn = i.Value.(*socketConn)
		if !conn.m_bConnect {
			continue
		}

		if mark != ConnMark_Def {
			if mark != conn.m_ConnMark {
				continue
			}
		}

		sendSuc = conn.Send(cmd, msg, msgSize)
	}

	_this.m_lock.Unlock()
	return sendSuc
}

//////

///////ClientSocket
type ClientSocket struct {
	m_Connect *socketConn
}

func (_this *ClientSocket) init() {
	_this.m_Connect = nil
}

//Start startClinet
func (_this *ClientSocket) Start(mark byte, Ip string, port string, recver *IClientRecver) bool {
	conn, err := net.DialTimeout("tcp", Ip+":"+port, 2*time.Second)
	if err != nil {
		fmt.Println("connect error:", err)
		return false
	}

	data := make([]byte, 1)
	data[0] = mark
	n, erro := conn.Write(data)
	if erro != nil {
		fmt.Println("connect error:", erro)
		return false
	}
	if n != 1 {
		return false
	}

	_this.m_Connect = new(socketConn)
	_this.m_Connect.init(conn, nil, recver)
	return true
}

//Send ClinetSendMsg
func (_this *ClientSocket) Send(cmd uint32, msg []byte, msgSize int) bool {
	if nil == _this.m_Connect {
		return false
	}
	return _this.m_Connect.Send(cmd, msg, msgSize)
}

////////
